### PR TITLE
Gyro and PID loop running separately

### PR DESCRIPTION
**Pull request by DzikuVx**
This is very experimental attempt, and very rough, proof-of-concept style, attempt to separate gyro and PID loop as it was discussed in #329 

PID loop is hardcoded at 500Hz and gyro loop at 2kHz. Until not I was able to determine that:
1. This is too much for F1 targets, CPU load was at 200%
2. Gyro at 1kHz and PID at 500Hz should be much more realistic on F1
3. But looks like it is working, PID loop was fed with data, output was generated. Did not flight tested it yet

Generally, it might have some potential... @digitalentity @martinbudden

### CODE REVIEW

**Comment by DzikuVx** [comment]
OK, if anybody is brave enough, it can be tried in flight... Made it more realistic : PID at 500Hz, gyro at 1kHz. I will try to take it to the air tomorrow, unfortunately do not have any F3 target ATM

@stronnag @theArchLadder who is brave enough?

**Comment by theArchLadder** [comment]
@DzikuVx I already run at 1khz on my F3, how fast can my gyro run on i2c and how fast can the pid loop run on an F3? Or if you don't know, how do i find out? I guess pid is the easy one, just watch cpu load, but what about the gyro?

**Comment by DzikuVx** [comment]
@theArchLadder No idea about F3, sorry. Right now I want to check if this kind of separation between loops is flyable or not. 1kHz/500Hz works on F1. 2kHz/500Hz does not. I think 1.5kHz/500Hz should be doable with i2c_overclock

cli commands: status and tasks is the way to find out.

**Comment by DzikuVx** [comment]
OK, 1kHz on gyro and 500Hz on PID is max what F1 target can do. Going above 1kHz on gyro give unstable looptime

**Comment by DzikuVx** [comment]
OK, Good news. Flight tested with SPRacingF3, 250 on 3S at 2kHz gyro loop and 500Hz PID loop. Quad never flew better. A lot credit goes for I-term changes, but besides that, wow, I'm very impressed. Somebody should confirm that. 

zrzut ekranu 2016-07-23 o 14 14 08

[2kHz_gyro_500Hz_PID_1.TXT.zip](https://github.com/iNavFlight/inav/files/379536/2kHz_gyro_500Hz_PID_1.TXT.zip)

F3 targets can start from `set gyro_sync_denom=4`, F1 `set gyro_sync_denom=8`
In both cases, PID looptime is independent and can be set as low as desired. 2000 should work for F1 and F3.
If gyro is I2C connected, I suggest `set i2c_overclock=ON`

**Comment by digitalentity** [comment]
@DzikuVx that's amazingly good news! Keep us posted with your findings!

**Comment by DzikuVx** [comment]
Short video of gyro at 2kHz and PID at 666Hz from today [https://youtu.be/iyylpFV69pw](https://youtu.be/iyylpFV69pw)
To compare: the same place and PID, but gyro in sync with PID loop at 666Hz from few days back [https://www.youtube.com/watch?v=aVpyRHTJghs](https://www.youtube.com/watch?v=aVpyRHTJghs)
For me there is a progress (minus my low skill)

**Comment by DzikuVx** [comment]
Looks like top limit for gyro readout right now in F3 targets over I2C is 2.6kHz. I will try to test it today

**Comment by DzikuVx** [comment]
2.6kHz/666Hz flies superb

**Comment by oleost** [comment]
@DzikuVx Could you dump the neccesary cli commands to get it 2.6kHZ/666hz?`

**Comment by DzikuVx** [comment]
```
set looptime=1500
set gyro_sync_denom=3
set i2c_overclock=ON
```

after that go to sensors tab, show "Debug" and verify that debug[0] is not jumping. Or run CLI `tasks` several times. Verify that frequency is somehat stable, does not jumps 2600->800 or smth like this.
Only F3 can do it, F1 only   `set gyro_sync_denom=8`

**Comment by oleost** [comment]
Unsure whats the dip in debug[0] is, it happend when I moved the quad, but I could not get it to dip anymore

image

image

**Comment by DzikuVx** [comment]
Should be flyable, start maybe with denom=4. That should be safer

**Comment by oleost** [comment]
Had an test flight with =3. Worked.

**Comment by oleost** [comment]
But I have not tuned my quad properly, so could not notice any difference. Just wanted to do a test flight.

**Comment by DzikuVx** [comment]
try 

```
set gyro_soft_lpf_hz = 90
set dterm_lpf_hz = 60
set yaw_lpf_hz = 45
```

That helped a lot

**Comment by oleost** [comment]
It's an 250 quad, shouldn't the filtering be set high?
1. jul. 2016 10.16 p.m. skrev "Paweł Spychalski" <notifications@github.com

File: src/main/scheduler/scheduler_tasks.c [hunk 1]
```diff
@@ -42,7 +42,14 @@ cfTask_t cfTasks[TASK_COUNT] = {
...
         .taskFunc = taskGyro,
         .desiredPeriod = 1000000 / 1000, //Run at 1000Hz
-        .staticPriority = TASK_PRIORITY_REALTIME
+        .staticPriority = TASK_PRIORITY_REALTIME,
+    },
+
+    [TASK_ACC] = {
+        .taskName = "ACC",
+        .taskFunc = taskAcc,
+        .desiredPeriod = 1000000 / 120, //Run at 120Hz, with LPF at 15Hz this is even too much
```
Comment (digitalentity) [inline src/main/scheduler/scheduler_tasks.c]: The problem is aliasing. If we have vibrations over 120Hz (which is in fact very common even for bigger machines) they will wrap down to 120Hz range - 121Hz will be wrapped down to 1Hz and pass though LPF no problem. We've been down that road before when scheduler was introduced.

Comment (DzikuVx) [inline src/main/scheduler/scheduler_tasks.c]: I realize that. I'm trying to get as much as possible in terms of performance ATM

Comment (digitalentity) [inline src/main/scheduler/scheduler_tasks.c]: Acro-only this should be fine, ANGLE might (and probably) will drift. Navigation will also be heavily affected. We shouldn't break ANGLE/Navigation while chasing performance, however for R&D I figure this is fine :smiley:

Comment (DzikuVx) [inline src/main/scheduler/scheduler_tasks.c]: I'm thinking how much is a threshold here. I remember the time when control loop was running at 285Hz and everyone was fine about that....

Comment (digitalentity) [inline src/main/scheduler/scheduler_tasks.c]: Back then quads were bigger, props we spinning slower and vibrations were of lower frequency....

Comment (DzikuVx) [inline src/main/scheduler/scheduler_tasks.c]: Personally I see not problem to have a switch/feature called, let's say "RACE" that will go to async acc to save some cpu resources for faster/better gyro processing

Comment (DzikuVx) [inline src/main/scheduler/scheduler_tasks.c]: I did some research, and acc task consumes the same cpu as PID task on F1 target

**Comment by DzikuVx** [comment]
OK, latest changes to acc will be moved into another branch not to complicate things too much. This can be tested without acc infuence

**Comment by martinbudden** [comment]
@DzikuVx , very interesting work. I certainly think this is the way to go. I don't have time to actively contribute in this area at the moment, but I'm following your changes with interest.

**Comment by DzikuVx** [comment]
PR replaced with #421

**Pull request closed without merging**